File: 133-clone-graph/133-clone-graph.scala
/**
 * Definition for a Node.
 * class Node(var _value: Int) {
 *   var value: Int = _value
 *   var neighbors: List[Node] = List()
 * }
 */

import scala.collection.mutable._

object Solution {
    def cloneGraph(graph: Node): Node = {
        if (graph != null) clone(graph, Map()) else null
    }
    
    def clone(graph: Node, map: Map[Int, Node]): Node = {
        map.get(graph.value).getOrElse {
            val node = new Node(graph.value)
            map += (node.value -> node)
            node.neighbors = graph.neighbors.map(clone(_, map))
            node
        }
    }
}

File: 133-clone-graph/133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    Map<Node, Node> cloneMap = new HashMap<>();
    
    public Node cloneGraph(Node node) {
        if (node == null) return null;
        if (cloneMap.containsKey(node)) return cloneMap.get(node);
        
        Node cloneNode = new Node(node.val);
        
        cloneMap.put(node, cloneNode);
        
        for(Node n : node.neighbors) {
            Node cloneNeighbor = cloneGraph(n);
            cloneNode.neighbors.add(cloneNeighbor);
        }
        
        return cloneNode;
    }
}